In [1]:
import gym
import numpy as np
import random
import time

#Q

In [2]:
def q_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate):
    q_table = np.zeros((env.observation_space.n, env.action_space.n))
    rewards_all = []
    for episode in range(num_episodes):
        state = env.reset()

        reward_episode = 0.0
        done = False
        epsilon = min_epsilon + (max_epsilon - min_epsilon) * np.exp(-epsilon_decay_rate*episode)
        for step in range(num_steps_per_episode):
            exploration = random.uniform(0,1)
            if exploration < epsilon:
                action = env.action_space.sample()
            else:
                action = np.argmax(q_table[state, :])

            next_state, reward, done, info = env.step(action)
            q_table[state, action] = q_table[state, action] * (1 - learning_rate) + learning_rate * (reward + gamma * np.max(q_table[next_state,:]))

            reward_episode += reward
            state = next_state

            if done:
                break
        rewards_all.append(reward_episode)
    print(f'Episode {episode} finished')
    return q_table, rewards_all

In [3]:
def play(env, q_table, render=False):
    state = env.reset()
    total_reward = 0
    steps = 0
    done = False
    while not done:
        action = np.argmax(q_table[state, :])
        next_state, reward, done, info = env.step(action)
        total_reward += reward
        steps += 1
        if render:
            env.render()
            time.sleep(0.2)
            if not done:
                display.clear_output(wait=True)
        state = next_state

    return (total_reward, steps)

In [4]:
def play_multiple_times(env, q_table, max_episodes):
    success = 0
    list_of_steps = []
    for i in range(max_episodes):
        total_reward, steps = play(env, q_table)

        if total_reward > 0:
            success += 1
            list_of_steps.append(steps)

    print(f'Number of successes: {success}/{max_episodes}')
    print(f'Average number of steps: {np.mean(list_of_steps)}')

## FrozenLake-v0

In [5]:
env = gym.make('FrozenLake-v0')

# Initialize Q-value table randomly
q_table = np.zeros((env.observation_space.n, env.action_space.n))

# Hyperparameters
gamma = 0.99
learning_rate = 0.1
max_epsilon = 1.0
min_epsilon = 0.01
epsilon_decay_rate = 0.005

num_episodes = 20000
num_steps_per_episode = 100



In [6]:
start = time.time()
q_table, rewards_all = q_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate)
end = time.time()
play_multiple_times(env, q_table, 1000)
print("Running time: ", end-start)

Episode 19999 finished
Number of successes: 741/1000
Average number of steps: 37.178137651821864
Running time:  19.070709943771362


## FrozenLake8x8-v0

In [19]:
env = gym.make('FrozenLake8x8-v0')

# Initialize Q-value table randomly
q_table = np.zeros((env.observation_space.n, env.action_space.n))

# Hyperparameters
gamma = 0.9
learning_rate = 0.8
max_epsilon = 1.0
min_epsilon = 0.001
epsilon_decay_rate = 0.00005

num_episodes = 250000
num_steps_per_episode = 400

In [20]:
start = time.time()
q_table, rewards_all = q_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate)
end = time.time()
play_multiple_times(env, q_table, 1000)
print("Running time: ", end-start)

Episode 249999 finished
Number of successes: 499/1000
Average number of steps: 91.7434869739479
Running time:  509.27175998687744


## Taxi-v3

In [9]:
env = gym.make('Taxi-v3')

# Initialize Q-value table randomly
q_table = np.zeros((env.observation_space.n, env.action_space.n))

# Hyperparameters
gamma = 0.99
learning_rate = 0.1
max_epsilon = 1.0
min_epsilon = 0.01
epsilon_decay_rate = 0.005

num_episodes = 20000
num_steps_per_episode = 100

q_table, rewards_all = q_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate)

Episode 19999 finished


In [10]:
start = time.time()
q_table, rewards_all = q_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate)
end = time.time()
play_multiple_times(env, q_table, 1000)
print("Running time: ", end-start)

Episode 19999 finished
Number of successes: 1000/1000
Average number of steps: 13.058
Running time:  9.554579496383667


#SARSA

In [25]:
def SARSA_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate):
    q_table = np.zeros((env.observation_space.n, env.action_space.n))
    rewards_all = []
    for episode in range(num_episodes):
        state = env.reset()

        reward_episode = 0.0
        done = False
        epsilon = min_epsilon + (max_epsilon - min_epsilon) * np.exp(-epsilon_decay_rate*episode)

        exploration = random.uniform(0,1)
        if exploration < epsilon:
            action = env.action_space.sample()
        else:
            action = np.argmax(q_table[state, :])

        for step in range(num_steps_per_episode):
            
            next_state, reward, done, info = env.step(action)
            
            exploration = random.uniform(0,1)
            if exploration < epsilon:
                next_action = env.action_space.sample()
            else:
                next_action = np.argmax(q_table[next_state, :])

            q_table[state, action] = q_table[state, action] * (1 - learning_rate) + learning_rate * (reward + gamma * q_table[next_state,next_action])


            reward_episode += reward
            action = next_action
            state = next_state

            if done:
                break
        rewards_all.append(reward_episode)
    print(f'Episode {episode} finished')
    return q_table, rewards_all

## FrozenLake-v0

In [26]:
env = gym.make('FrozenLake-v0')

# Initialize Q-value table randomly
q_table = np.zeros((env.observation_space.n, env.action_space.n))

# Hyperparameters
gamma = 0.99
learning_rate = 0.1
max_epsilon = 1.0
min_epsilon = 0.01
epsilon_decay_rate = 0.005

num_episodes = 20000
num_steps_per_episode = 100



In [27]:
start = time.time()
q_table, rewards_all = SARSA_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate)
end = time.time()
play_multiple_times(env, q_table, 1000)
print("Running time: ", end-start)

Episode 19999 finished
Number of successes: 629/1000
Average number of steps: 37.20826709062003
Running time:  12.387576818466187


## FrozenLake8x8-v0

In [23]:
env = gym.make('FrozenLake8x8-v0')

# Initialize Q-value table randomly
q_table = np.zeros((env.observation_space.n, env.action_space.n))

# Hyperparameters
gamma = 0.9
learning_rate = 0.8
max_epsilon = 1.0
min_epsilon = 0.001
epsilon_decay_rate = 0.00005

num_episodes = 250000
num_steps_per_episode = 400

# q_table, rewards_all = SARSA_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate)

In [24]:
start = time.time()
q_table, rewards_all = SARSA_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate)
end = time.time()
play_multiple_times(env, q_table, 1000)
print("Running time: ", end-start)

Episode 249999 finished
Number of successes: 729/1000
Average number of steps: 86.58710562414267
Running time:  313.45714259147644


## Taxi-v3

In [17]:
env = gym.make('Taxi-v3')

# Initialize Q-value table randomly
q_table = np.zeros((env.observation_space.n, env.action_space.n))

# Hyperparameters
gamma = 0.99
learning_rate = 0.1
max_epsilon = 1.0
min_epsilon = 0.01
epsilon_decay_rate = 0.005

num_episodes = 20000
num_steps_per_episode = 100


Episode 19999 finished


In [18]:
start = time.time()
q_table, rewards_all = SARSA_learning(env, num_episodes, num_steps_per_episode, learning_rate, gamma, max_epsilon, min_epsilon, epsilon_decay_rate)
end = time.time()
play_multiple_times(env, q_table, 1000)
print("Running time: ", end-start)

Episode 19999 finished
Number of successes: 1000/1000
Average number of steps: 13.094
Running time:  5.513466835021973


# Nhận xét

FrozenLake-v0 | Q-learning  | SARSA 
-------------------|-------------------|------------------
Running time |19.070709943771362 |12.387576818466187
Number of successes       | 741/1000 |629/1000
Average number of steps       | 37.178137651821864| 37.20826709062003


FrozenLake8x8-v0 | Q-learning  | SARSA
-------------------|-------------------|------------------
Running time |509.27175998687744 | 313.45714259147644
Number of successes       | 499/1000 |729/1000
Average number of steps       |91.7434869739479 | 86.58710562414267


Taxi-v3 | Q-learning  | SARSA
-------------------|-------------------|------------------
Running time |9.554579496383667 |5.513466835021973
Number of successes       | 1000/1000 |1000/1000
Average number of steps       | 13.058 | 13.094


* Thời gian tìm lời giải của 2 thuật toán có sự chênh lệch lớn, trong cả 3 trường hợp SARSA learing có thời gian tìm lời giải chỉ xấp xỉ 3/5 Q learning
*  Số bước giải trung bình ở 2 thuật toán không quá chênh lệch
* Q learning có tỉ lệ thắng cao hơn ở map FrozenLake-v0 , còn SARSA có tỉ lệ thắng cao hơn ở map FronzenLake8x8-v0